ssl/handshake_client.cc

```diff
@@ -232,9 +259,23 @@ static bool ssl_write_client_cipher_list(const SSL_HANDSHAKE *hs, CBB *out,
     return false;
   }
 
-  // Add TLS 1.3 ciphers. Order ChaCha20-Poly1305 relative to AES-GCM based on
-  // hardware support.
-  if (hs->max_version >= TLS1_3_VERSION) {
+  // Add TLS 1.3 ciphers.
+  if (hs->max_version >= TLS1_3_VERSION && ssl->ctx->tls13_cipher_list) {
+    // Use the configured TLSv1.3 ciphers list.
+    STACK_OF(SSL_CIPHER) *ciphers = ssl->ctx->tls13_cipher_list->ciphers.get();
+    bool any_enabled = false;
+    if (!collect_cipher_protocol_ids(ciphers, &child, mask_k,
```

Comment:
Why should these extra conditions apply just because you restrict the list of tls 1.3 cipher suites? Why wouldn't they apply regardless?

Reply:
Isn't there more restrictions applied as well? thinking about:
```
    // Skip disabled ciphers
    if ((cipher->algorithm_mkey & mask_k) ||
        (cipher->algorithm_auth & mask_a)) {
      continue;
    }
```

Reply:
No. This is just wrapper function of [existing code](https://github.com/awslabs/aws-lc/blob/123eaaef26abc278f53ae338e9c758eb01c70b08/ssl/handshake_client.cc#L254-L267). It's wrapped to be reused.

Reply:
Yes. I was saying that this now applies to TLS 1.3 cipher suite selection - if you have restricted the list - but if you have not restricted the list, the previous behaviour is used. No?

Reply:
I just see a discrepancy:
* if a customer set of TLS 1.3 ciphers is set, we take a different code path that could (potentially) apply some restrictions.
* if build-in set, we don't do that.

So, conceptually we now have something like this (with reference to https://bugs.chromium.org/p/boringssl/issues/detail?id=110)?
* if a custom set of TLS 1.3 ciphers is set, we somehow try to "re-attach" to the existing cipher language.
* but if no custom set, we keep the detach behaviour.

I just find it rather complex.

Reply:
> if a customer set of TLS 1.3 ciphers is set, we take a different code path that could (potentially) apply some restrictions.

Not sure what the concern is. Below are confirmed.
1. The restriction does not apply to TLSv1.3 ciphers.
2. In future, the tests can tell if unexpected restrictions are applied.

> if a custom set of TLS 1.3 ciphers is set, we somehow try to "re-attach" to the existing cipher language.

Not "re-attach" but only choose the selected TLSv1.3 ciphers. This is the expected functionality of the API `|SSL_CTX_set_ciphersuites|`.

> I just find it rather complex.

What's the alternative?